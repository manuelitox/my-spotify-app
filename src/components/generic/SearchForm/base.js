import React from 'react'
import PropTypes from 'prop-types'

import StyledInput from './stylesInput'
import LoadingMessage from './loading'
import SuggestionsList from './suggestionsList'
import StyledSearchForm from './styles'
import ArtistMatchesCounter from './artistsCounter'

const SearchForm = ({ 
  data,
  token,
  isOpen,
  loading,
  onChange,
  placeholder,
  openSuggestions,
  closeSuggestions
}) => (
  <StyledSearchForm>
    <StyledInput 
      id="search"
      type="search"
      isOpen={ isOpen }
      onClick={ openSuggestions }
      onChange={ onChange }
      placeholder={ placeholder }
      autocomplete="off" />
    <LoadingMessage loading={ loading } />
    <ArtistMatchesCounter isOpen={ isOpen } loading={ loading } numberOfArtists={ data.length } />
    <SuggestionsList isOpen={ isOpen } numberOfArtists={ data.length } data={ data } closeSuggestions={ closeSuggestions } token={ token } />
  </StyledSearchForm>
)  

SearchForm.propTypes = {
  data: PropTypes.array.isRequired,
  token: PropTypes.string,
  isOpen: PropTypes.bool,
  loading: PropTypes.bool,
  onChange: PropTypes.func.isRequired,
  placeholder: PropTypes.string.isRequired,
  openSuggestions: PropTypes.func.isRequired,
  closeSuggestions: PropTypes.func.isRequired
}

export default SearchForm