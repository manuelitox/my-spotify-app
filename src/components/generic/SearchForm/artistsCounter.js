import React from 'react'
import PropTypes from 'prop-types'

import StyledMessage from './stylesMessage'

const ArtistMatchesCounter = ({ 
  isOpen,
  loading,
  numberOfArtists
}) => (
  !loading && isOpen ? <StyledMessage>{ numberOfArtists } matches</StyledMessage> : null
)

ArtistMatchesCounter.propTypes = {
  isOpen: PropTypes.bool,
  loading: PropTypes.bool,
  numberOfArtists: PropTypes.number
}

export default ArtistMatchesCounter