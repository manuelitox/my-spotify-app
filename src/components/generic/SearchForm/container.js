import React, { Component } from 'react'
import onClickOutside from 'react-onclickoutside'

import SearchForm from './base'

export class SearchFormContainer extends Component {
  state = {
    isOpen: false,
    loading: false,
    suggestions: []
  }

  handleClickOutside = evt => {
    this.setState({ isOpen: false })
  }

  openSuggestions = () => {
    const { suggestions } = this.state
    if ( suggestions.length >= 1 ) this.setState({ isOpen: true })
  }

  onChange = e => {
    console.log(e.target.value)
    this.setState({ 
      isOpen: true,
      loading: true 
    })
    setTimeout(() => {
      this.setState({ 
        suggestions: [
          { id: '1', name: 'Alicia Keys' },
          { id: '2', name: 'Queen' },
          { id: '3', name: 'John Mayer' },
          { id: '4', name: 'ACDC' },
          { id: '5', name: 'Blink 182' },
          { id: '6', name: 'Bon Jovi' },
          { id: '7', name: '2pac' }          
        ],
        loading: false
    })
    }, 1000);
  }

  render () {
    return (
      <SearchForm 
        onChange={ this.onChange } 
        openSuggestions={ this.openSuggestions } 
        { ...this.state } />
    )
  }
}

export default onClickOutside(SearchFormContainer)