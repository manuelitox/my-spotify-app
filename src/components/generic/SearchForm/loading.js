import React from 'react'
import PropTypes from 'prop-types'

import StyledMessage from './stylesMessage'

const LoadingMessage = ({ loading }) => (
  loading ? <StyledMessage>loading</StyledMessage> : null
)

LoadingMessage.propTypes = {
  loading: PropTypes.bool
}

export default LoadingMessage