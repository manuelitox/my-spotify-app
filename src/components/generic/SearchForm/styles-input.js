import styled from 'styled-components'

const StyledInput = styled.input`
  appearance: none;
  border: 0;
  border-radius: 4px;
  padding: 20px;
`

export default StyledInput