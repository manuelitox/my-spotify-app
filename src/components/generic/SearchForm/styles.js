import styled from 'styled-components'

const StyledSearchForm = styled.div`
  margin: 20px auto 0;
  position: relative;
  width: 560px;
  @media only screen and (max-width: ${ props => props.theme.size.small }) {
    width: 100%;
  }
`

export default StyledSearchForm