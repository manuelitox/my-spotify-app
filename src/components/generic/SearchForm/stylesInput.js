import styled from 'styled-components'
import PropTypes from 'prop-types';

const StyledInput = styled.input`
  appearance: none;
  border: 0;
  border-radius: ${ props => props.isOpen ? '4px 4px 0 0' : '4px' };
  padding: 20px;
  &:focus {
    outline: 0;
  }
`

StyledInput.propTypes = {
  isOpen: PropTypes.bool
}

export default StyledInput