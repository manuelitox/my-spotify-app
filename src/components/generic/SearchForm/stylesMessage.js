import styled from 'styled-components'

const StyledMessage = styled.div`
  background-color: ${ props => props.theme.searchSpan };
  border-radius: 4px;
  color: ${ props => props.theme.offWhite };
  font-size: 1rem;
  font-weight: 600;
  right: 14px;
  top: 16px;
  padding: 8px;
  position: absolute;
`

export default StyledMessage