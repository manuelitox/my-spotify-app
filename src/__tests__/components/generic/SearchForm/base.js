import React from 'react'
import { shallow } from 'enzyme'

import SearchForm from 'components/generic/SearchForm/base'
import SuggestionsList from 'components/generic/SearchForm/suggestionsList'

describe('Component | SearchForm', () => {
  const SUGGESTIONS = [
    { id: '1', name: 'Alicia Keys' },
    { id: '2', name: 'Queen' },
    { id: '3', name: 'John Mayer' },
    { id: '4', name: 'ACDC' },
    { id: '5', name: 'Blink 182' },
    { id: '6', name: 'Bon Jovi' },
    { id: '7', name: '2pac' }          
  ]
  let component 
  beforeEach( () => {
    component = shallow(
      <SearchForm  
        data={ SUGGESTIONS }
        token="access_token123123"
        onChange={ jest.fn() }
        placeholder="search an artist"
        openSuggestions={ jest.fn() }
        closeSuggestions={ jest.fn() } />
    )
  })
  it('should render a component', () => {
    expect(component).toMatchSnapshot()
  })
})