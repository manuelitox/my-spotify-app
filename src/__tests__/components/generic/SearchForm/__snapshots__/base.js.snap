// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component | SearchForm should render a component 1`] = `
<styled.div>
  <styled.input
    id="search"
    onChange={[MockFunction]}
    onClick={[MockFunction]}
    placeholder="search an artist"
    type="search"
  />
  <LoadingMessage />
  <ArtistMatchesCounter
    numberOfArtists={7}
  />
  <SuggestionsList
    closeSuggestions={[MockFunction]}
    data={
      Array [
        Object {
          "id": "1",
          "name": "Alicia Keys",
        },
        Object {
          "id": "2",
          "name": "Queen",
        },
        Object {
          "id": "3",
          "name": "John Mayer",
        },
        Object {
          "id": "4",
          "name": "ACDC",
        },
        Object {
          "id": "5",
          "name": "Blink 182",
        },
        Object {
          "id": "6",
          "name": "Bon Jovi",
        },
        Object {
          "id": "7",
          "name": "2pac",
        },
      ]
    }
    numberOfArtists={7}
  />
</styled.div>
`;
